src/common/epoll_event.h

```diff
@@ -0,0 +1,175 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Intel <mahati.chamarthy@intel.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+#ifndef CEPH_LIBRADOS_EPOLLEVENT_H
+#define CEPH_LIBRADOS_EPOLLEVENT_H
+
+#include <unistd.h>
+#include <sys/epoll.h>
+#include "include/Context.h"
+#include "common/errno.h"
+#include "common/Mutex.h"
+#include "common/dout.h" 
+
+class CephContext;
+
+//int n_event = 5000;
+
+class EpollEvent{
+  int ep_fd;
+  int size;
+  struct epoll_event *events;
+  CephContext *cct;
+
+ public:
+  typedef void (*aio_callback_t)(void *arg);
+
+  EpollEvent() {}
+  //EpollEvent& operator=(const EpollEvent& e);
+  ~EpollEvent() {
+     if (ep_fd != -1)
+       close(ep_fd);
+
+     if (events)
+       free(events);
+  }
+
+  struct FiredAIOEvent {
+    int fd;
+  };
+
+  struct AIOEvent {
+    aio_callback_t complete_cb;
+    void *complete_arg;
+  };
+
+  vector<AIOEvent> aio_events;
+
+  //int n_event = 5000;
+
+  AIOEvent *_get_aio_event(int fd) {
+    assert(fd < 10000);
+    return &aio_events[fd];
+  }
+
+  int init(){
+    size = 5000;
+    events = (struct epoll_event*)malloc(sizeof(struct epoll_event)*size);
```

Comment:
@dillaman AsyncMessenger uses a fixed number of events (5000). But I don't think that number would suffice in this case, if every I/O call has an eventfd. I'm yet to think of a way to scale this dynamically. Any thoughts?

Plus, this PR ties up memory allocation within ThreadPoolSingleton - it isn't yet designed to malloc/free properly everytime ThreadPool is started and stopped. I'm thinking if EpollEvent should be tied up to ImageCtx. Although that way all the API calls of ThreadPool would need to be accommodated. Thoughts?

Reply:
Isn't this just the max number of events to return w/ each ```epoll_wait``` call? I would actually think you would want to keep this pretty small and perhaps even just thread-local storage. It's not like you will want a single thread to attempt to process 5000 events and then leave other threads idle.

Reply:
That's correct. Though at the moment epoll fd (epoll_create) is created per ThreadPool rather than per Thread, hence the higher number.